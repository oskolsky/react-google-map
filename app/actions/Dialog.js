export const OPEN_DIALOG = 'OPEN_DIALOG'
export const CLOSE_DIALOG = 'CLOSE_DIALOG'

export const openDialog = (content) => {
    return {
        type: OPEN_DIALOG,
        content: content,
    }
}

export const closeDialog = () => {
    return {
        type: CLOSE_DIALOG,
    }
}
