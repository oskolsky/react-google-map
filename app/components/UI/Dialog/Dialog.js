import React from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'

import { KEY_CODE } from 'helpers/keycodes'

import styles from './Dialog.scss'

class Dialog extends React.Component {
    static propTypes = {
        content: PropTypes.node.isRequired,
        isOpened: PropTypes.bool.isRequired,
        onCloseDialog: PropTypes.func.isRequired,
    }

    componentWillMount() {
        window.addEventListener('keydown', this.handleKeyDown, true)
        window.addEventListener('click', this.handleOutsideClick, true)
    }

    componentWillUnmount() {
        window.removeEventListener('keydown', this.handleKeyDown, true)
        window.removeEventListener('click', this.handleOutsideClick, true)
    }

    handleKeyDown = (e) => {
        if (this.props.isOpened) {
            if (e.keyCode === KEY_CODE.ESC) {
                this.props.onCloseDialog()
                e.stopImmediatePropagation()
            }
        }
    }

    handleOutsideClick = (e) => {
        if (this.props.isOpened && !this._dialog.contains(e.target)) {
            this.props.onCloseDialog()
        }
    }

    render() {
        const classNameBase = classNames(styles.base, {
            [styles.isOpened]: this.props.isOpened,
        })

        return (
            <div className={classNameBase}>
                <div className={styles.dialog} ref={(c) => this._dialog = c}>
                    <div className={styles.close} onClick={this.props.onCloseDialog}>Закрыть</div>
                    <div className={styles.body}>
                        {this.props.content}
                    </div>
                </div>
            </div>
        )
    }
}

export default Dialog
