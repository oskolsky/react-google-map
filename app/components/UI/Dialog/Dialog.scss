@import '~stylesheets/core';

.base {
    position: fixed;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
    z-index: map-get($zindex, dialog);
    display: none;
    flex-direction: column;
    align-items: center;
    justify-content: center;

    &::after {
        position: absolute;
        top: 0;
        right: 0;
        bottom: 0;
        left: 0;
        z-index: 1;
        background: rgba(0, 0, 0, .8);
        content: '';
    }

    &.isOpened {
        display: flex;
    }
}

.dialog {
    position: relative;
    z-index: 2;
    width: 500px;
    height: 500px;
    padding: 20px;
    background: #fff;
}

.close {
    position: absolute;
    top: -52px;
    right: 9px;
    color: #fff;
}
