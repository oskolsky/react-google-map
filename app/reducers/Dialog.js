import {
    OPEN_DIALOG,
    CLOSE_DIALOG,
} from 'actions/Dialog'

const initialState = {
    content: '',
    isOpened: false,
}

export const dialog = (state = initialState, action) => {
    switch (action.type) {
        case OPEN_DIALOG: {
            return {
                ...state,
                content: action.content,
                isOpened: true,
            }
        }
        case CLOSE_DIALOG: {
            return {...state, content: '', isOpened: false }
        }
        default:
            return state
    }
}
