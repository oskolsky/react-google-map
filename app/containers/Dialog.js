import { connect } from 'react-redux'

import { Dialog } from 'components'
import { closeDialog } from 'actions/Dialog'

const mapStateToProps = (state) => {
    return {
        content: state.dialog.content,
        isOpened: state.dialog.isOpened,
    }
}

const mapDispatchToProps = {
    onCloseDialog: closeDialog,
}

export default connect(mapStateToProps, mapDispatchToProps)(Dialog)
