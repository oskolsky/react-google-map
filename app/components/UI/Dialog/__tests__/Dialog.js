import React from 'react'
import { shallow } from 'enzyme'

import Dialog from '../Dialog'

function setup(options) {
    const props = {
        content: 'This is Dialog',
        isOpened: false,
        onCloseDialog: jest.fn(),
    }

    const enzymeWrapper = shallow(
        <Dialog {...props} {...options} />
    )

    return {
        props,
        enzymeWrapper,
    }
}

let enzymeWrapper = null

describe('Dialog specification', () => {
    afterEach(() => {
        enzymeWrapper.unmount()
    })

    it('renders correctly', () => {
        enzymeWrapper = setup().enzymeWrapper

        expect(enzymeWrapper).toMatchSnapshot()
    })

    it('is opened renders correctly', () => {
        enzymeWrapper = setup({ isOpened: true }).enzymeWrapper

        expect(enzymeWrapper).toMatchSnapshot()
    })
})
