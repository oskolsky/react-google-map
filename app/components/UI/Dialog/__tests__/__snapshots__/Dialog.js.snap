// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dialog specification is opened renders correctly 1`] = `
<div
  className="base isOpened"
>
  <div
    className="dialog"
  >
    <div
      className="close"
      onClick={[MockFunction]}
    >
      Закрыть
    </div>
    <div
      className="body"
    >
      This is Dialog
    </div>
  </div>
</div>
`;

exports[`Dialog specification renders correctly 1`] = `
<div
  className="base"
>
  <div
    className="dialog"
  >
    <div
      className="close"
      onClick={[MockFunction]}
    >
      Закрыть
    </div>
    <div
      className="body"
    >
      This is Dialog
    </div>
  </div>
</div>
`;
